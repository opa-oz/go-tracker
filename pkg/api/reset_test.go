package api_test

import (
	"github.com/gkampitakis/go-snaps/snaps"
	"net/http"
	"net/http/httptest"
	"testing"
	"tracker/pkg/api"
	"tracker/pkg/middlewares"

	"github.com/gin-gonic/gin"
	"github.com/go-playground/assert/v2"
	"github.com/go-redis/redismock/v9"
)

func TestResetRoute(t *testing.T) {
	r := gin.Default()

	rdb, mock := redismock.NewClientMock()

	r.Use(middlewares.RedisMiddleware(rdb))
	r.GET("/reset/:name", api.Reset)
	r.GET("/r/:name", api.Reset)

	t.Run("if keys is not set, reset to 0", func(t *testing.T) {
		mock.ExpectSet("test", int64(0), 0).SetVal("0")
		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/reset/test", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"value":0}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("if key is set, reset to 0", func(t *testing.T) {
		mock.ExpectSet("test", int64(0), 0).SetVal("0")

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/reset/test", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"value":0}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("if init is set, reset to init value", func(t *testing.T) {
		mock.ExpectSet("test", int64(50), 0).SetVal("0")

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/r/test?init=50", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"value":50}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})
}
