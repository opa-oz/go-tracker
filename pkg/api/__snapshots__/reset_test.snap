
[TestResetRoute/if_keys_is_not_set,_reset_to_0 - 1]
{"value":0}
---

[TestResetRoute/if_key_is_set,_reset_to_0 - 1]
{"value":0}
---

[TestResetRoute/if_init_is_set,_reset_to_init_value - 1]
{"value":50}
---
